### PR TITLE
APP-464: [C++ SDK] GPIO control slave application 

### PR DESCRIPTION
**Pull request by Slash32**
Re-base of #563 onto master.
Also fixed build issue.

### CODE REVIEW

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Oldnickus** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by vadimol** [comment]
@kaa-leeroy test this please

**Comment by vadimol** [comment]
@kaa-leeroy test this please

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        nsRemoteControlECF::GpioStatus status;
+        
+        status.status = false;
+        for (int n = 0; n < NUM_GPIO_LEDS; n++) {
+            status.id = n;
+            leds.push_back(status);
+        }
+    }
+
+    void onEvent(const nsRemoteControlECF::DeviceInfoRequest& event, const std::string& source)
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp]: Add `override`

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+#include <kaa/Kaa.hpp>
+#include <kaa/event/registration/IUserAttachCallback.hpp>
+#include <kaa/event/IFetchEventListeners.hpp>
+#include <kaa/event/gen/EventFamilyFactory.hpp>
+#include <kaa/event/gen/RemoteControlECF.hpp>
+
+using namespace kaa;
+
+#define NUM_GPIO_LEDS 4
+
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp:32]: Please remove this line

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            }
+        }
+    }
+    
+    void _printLedStatus()
+    {
+        std::cout << "GPIO LED status: ";
+        
+        for (auto &led : leds) {
+            std::cout << led.status;
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp:89]: How about newline or space character between led statuses?

File: gpiocontrol/source/cpp/src/KaaDemo.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    /*
+     * Run the Kaa endpoint.
+     */
+    kaaClient->start();
+
+    std::cout << "GPIO demo started" << std::endl;
+
+    while (true) {
+        std::this_thread::sleep_for(std::chrono::seconds(1));
```
Comment (RostakaGmfun) [inline gpiocontrol/source/cpp/src/KaaDemo.cpp:116]: Sleeping is not required here. The `while(true);` endless loop is sufficient

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 2 above)
Comment (vadimol) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:32]: Please, explain your point in your comments.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 3 above)
Comment (vadimol) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:89]: I think it is ok. We are trying to unify all of the applications. Currently, the GPIO slave demo on C has the same behavior. The format of the output can be reviewed in a separate Jira.

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 4 above)
Comment (vadimol) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:116]: Endless loops cause a high CPU usage. The sleep call allows yielding the CPU to other applications

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 3 above)
Comment (RostakaGmfun) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:89]: Ok, got it

File: gpiocontrol/source/cpp/src/KaaDemo.cpp (hunk 2 above)
Comment (RostakaGmfun) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:32]: It is empty

Comment (vadimol) [reply gpiocontrol/source/cpp/src/KaaDemo.cpp:32]: ah, ok. I though that you are talking about the line with "define".

**Comment by vadimol** [comment]
@kaa-leeroy test this please

**Comment by vadimol** [comment]
@kaa-leeroy test this please

**Comment by vadimol** [comment]
@RostakaGmfun , all comments were addressed.

**Review by RostakaGmfun — APPROVED** [review]

**Review by vadimol — APPROVED** [review]

**Pull request merged**